receiver/otlpreceiver/otlp.go

```diff
@@ -191,40 +187,160 @@ func (r *otlpReceiver) Shutdown(ctx context.Context) error {
+			handleHttpTraces(resp, req, pbContentType, r.traceReceiver, tracesPbUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		r.httpMux.HandleFunc("/v1/traces", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpTraces(resp, req, jsonContentType, r.traceReceiver, tracesJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
+		// For backwards compatibility see https://github.com/open-telemetry/opentelemetry-collector/issues/1968
+		r.httpMux.HandleFunc("/v1/trace", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpTraces(resp, req, jsonContentType, r.traceReceiver, tracesJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
 	}
 	return nil
 }
 
-func (r *otlpReceiver) registerMetricsConsumer(ctx context.Context, mc consumer.Metrics) error {
+var metricsPbUnmarshaler = otlp.NewProtobufMetricsUnmarshaler()
+var metricsJSONUnmarshaler = otlp.NewJSONMetricsUnmarshaler()
+
+func (r *otlpReceiver) registerMetricsConsumer(mc consumer.Metrics) error {
 	if mc == nil {
 		return componenterror.ErrNilNextConsumer
 	}
 	r.metricsReceiver = metrics.New(r.cfg.ID(), mc)
-	if r.gatewayMux != nil {
-		return collectormetrics.RegisterMetricsServiceHandlerServer(ctx, r.gatewayMux, r.metricsReceiver)
+	if r.httpMux != nil {
+		r.httpMux.HandleFunc("/v1/metrics", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpMetrics(resp, req, pbContentType, r.metricsReceiver, metricsPbUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		r.httpMux.HandleFunc("/v1/metrics", func(resp http.ResponseWriter, req *http.Request) {
+			handleHttpMetrics(resp, req, jsonContentType, r.metricsReceiver, metricsJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
 	}
 	return nil
 }
 
-func (r *otlpReceiver) registerLogsConsumer(ctx context.Context, tc consumer.Logs) error {
-	if tc == nil {
+var logsPbUnmarshaler = otlp.NewProtobufLogsUnmarshaler()
+var logsJSONUnmarshaler = otlp.NewJSONLogsUnmarshaler()
+
+func (r *otlpReceiver) registerLogsConsumer(lc consumer.Logs) error {
+	if lc == nil {
 		return componenterror.ErrNilNextConsumer
 	}
-	r.logReceiver = logs.New(r.cfg.ID(), tc)
-	if r.gatewayMux != nil {
-		return collectorlog.RegisterLogsServiceHandlerServer(ctx, r.gatewayMux, r.logReceiver)
+	r.logReceiver = logs.New(r.cfg.ID(), lc)
+	if r.httpMux != nil {
+		r.httpMux.HandleFunc("/v1/logs", func(w http.ResponseWriter, req *http.Request) {
+			handleHttpLogs(w, req, pbContentType, r.logReceiver, logsPbUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", pbContentType)
+		r.httpMux.HandleFunc("/v1/logs", func(w http.ResponseWriter, req *http.Request) {
+			handleHttpLogs(w, req, jsonContentType, r.logReceiver, logsJSONUnmarshaler)
+		}).Methods(http.MethodPost).Headers("Content-Type", jsonContentType)
 	}
 	return nil
 }
+
+func handleHttpTraces(
```

Comment:
handleHTTPTraces

Btw, are there any non-HTTP handlers? We can simply call this handleTraces.

Reply:
grpc